### PR TITLE
Backport security audit fixes (V01-V07), snappy gRPC fix, and stream push panic fix to release-1.21

### PR DESCRIPTION
**Pull request by friedrichg**
## Summary

Prepares **v1.21.1-rc.0** by backporting all seven Quarkslab security-audit fixes (V01-V07), PR #7459 (snappy gRPC panic), and PR #7541 (stream push panic) onto the `release-1.21` branch.

## Backport status

| # | PR | Finding | Conflicts |
|---|---|---|---|
| 1 | #7473 | V03 config leakage | CHANGELOG only |
| 2 | #7474 | V07 gossip integrity | none |
| 3 | #7475 | V01 tenant impersonation | CHANGELOG only |
| 4 | #7512 | V02 stored XSS | CHANGELOG only |
| 5 | #7515 | V04 gzip DoS | none |
| 6 | #7518 | V06 gossip TCP limits | CHANGELOG only |
| 7 | #7570 | V05 histogram allocation | CHANGELOG + 4 code files |
| 8 | #7459 | snappy gRPC panic | CHANGELOG + workflow |
| 9 | #7541 | stream push slice-out-of-range panic | CHANGELOG only |

Plus a final commit bumping `VERSION` to `1.21.1-rc.0` per `RELEASE.md`.

## V05 (#7570) conflict resolution notes

- **CHANGELOG**: reclassified the upstream `[ENHANCEMENT]` line as `[BUGFIX]` to match the patch-release tone. All nine new entries appear under `## 1.21.1 in progress`.
- **integration/remote_write_v2_test.go**: dropped three incoming tests (`TestIngest_StartTimestamp`, `TestIngest_CreatedTimestampFallback`, `TestIngest_StartAndCreatedTimestampIgnoredWhenDisabled`); they belong to #7371 which isn't on release-1.21.
- **pkg/cortexpb/cortex.pb.go**: took the incoming `fileDescriptor` blob; it matches the cherry-picked `.proto`.
- **pkg/cortexpb/timeseriesv2_test.go**: kept the `Histograms []Histogram` -> `[]WrappedHistogram` swap and the `MetadataV2` reuse-clearing assertions; dropped the `CreatedTimestamp` setup + `assert.Zero` because `ReuseTimeseriesV2` on release-1.21 doesn't clear `CreatedTimestamp`, so the upstream assertion would fail. The `MetadataV2` field exists on release-1.21 and is reset by `ReuseTimeseriesV2`, so its coverage is preserved.
- **pkg/ingester/ingester_test.go**: dropped `TestIngester_Push_StartTimestamp` and `TestIngester_Push_StartTimestampAppendFailureMetrics`; both depend on `StartTimestamp`.
- **pkg/util/push/push_test.go**: dropped `Test_convertV2RequestToV1_PreservesStartTimestamp`; same reason.

## Test plan

- [x] `go build ./...`
- [x] `go test -timeout 2400s -tags "netgo slicelabels" ./...`
- [x] CI green on this PR
- [ ] Sanity test in a staging cluster after rc.0 is tagged

### CODE REVIEW

**Comment by SungJin1212** [comment]
@friedrichg 
Can you add this PR https://github.com/cortexproject/cortex/pull/7541?

**Review by sandy2008 — COMMENTED** [review]
Took a careful pass over the backport — diffed each cherry-pick against its upstream PR and built + vetted the branch under `-tags 'netgo slicelabels'`. Overall this is a clean, faithful backport: every production hunk for V01–V07 + the snappy fix is byte-identical to upstream, the `[]Histogram` → `[]WrappedHistogram` swap and the regenerated `cortex.pb.go` build green, and the dropped tests are the right ones (they depend on `StartTimestamp`/`CreatedTimestamp` from #7371, which isn't on 1.21).

A few small things inline — mostly test-quality (a couple of the security tests currently pass without actually exercising what they guard), one CHANGELOG/rc nit, and one empty-signing-key edge case carried over from #7475 (flagging for a follow-up, not a blocker). Thanks for putting this together!

File: pkg/util/flagext/secretstringslicecsv.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	return strings.Join(v.values, ",")
+}
+
+// Set implements flag.Value
+func (v *SecretStringSliceCSV) Set(s string) error {
+	if s == "" {
+		v.values = nil
+		return nil
+	}
+	v.values = strings.Split(s, ",")
```
Comment (sandy2008) [inline pkg/util/flagext/secretstringslicecsv.go:23]: `strings.Split` keeps empty entries, so a stray/trailing comma in `-distributor.sign-write-requests-keys` (e.g. `newkey,`, or a fat-fingered `newkey,oldkey`) leaves a `""` in the key list. The ingester's `NewStreamSigningServerInterceptor(keys...)` then accepts the stream when `ConstantTimeCompare(sig, computeStreamHMAC("", orgID)) == 1` — and `orgID` is the worker name (`ingester-<addr>-stream-push-worker-<N>`, documented just above), which isn't a secret. So an empty key collapses to `HMAC-SHA256("", <guessable>)`, i.e. forgeable with no secret → V01 stream auth silently downgrades to no-auth. The distributor keeps signing with `keys[0]`, so legit traffic still flows and the misconfig is invisible.

This is verbatim from #7475 (not introduced by the backport), so not blocking the rc — but since V01 exists to close tenant impersonation, might be worth hardening here and forward-porting: in `Set()`, `TrimSpace` each entry and drop/reject empties (error if nothing valid is left). A couple of cases like `newkey,` and `a,,b` would lock it in.

File: integration/ingester_stream_push_test.go [hunk 2]
```diff
@@ -116,6 +116,101 @@ func TestIngesterStreamPushConnection(t *testing.T) {
...
+		e2e.WaitMissingMetrics))
+
+	now := time.Now()
+	client, err := e2ecortex.NewClient(distributor.HTTPEndpoint(), "", "", "", userID)
+	require.NoError(t, err)
+
+	for i := 0; i < 3; i++ {
+		series, _ := generateSeries(fmt.Sprintf("test_signing_mismatch_%d", i), now)
+		res, err := client.Push(series)
+		if err == nil {
```
Comment (sandy2008) [inline integration/ingester_stream_push_test.go:207]: This assertion is effectively a no-op: `require.NotEqual(200, ...)` only runs when `err == nil`, but a signature mismatch over the stream will typically surface as a non-nil `client.Push` error → the loop then asserts nothing, and the test would still pass even if stream signing were completely broken. There's also no positive check that the rejection is signature-related.

Could we make it deterministic — `require.NoError(t, err)`, then assert a non-200 status, and ideally that the body/error carries `signature mismatch` (the `ErrSignatureMismatch` from `signing_handler.go`)? (Carried over from #7475, but it's a security test, so the false-confidence is the concerning part.)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## 1.21.1 / 2026-06-01
```
Comment (sandy2008) [inline CHANGELOG.md]: nit: this is the `rc.0` PR (`VERSION` = `1.21.1-rc.0`), so per `RELEASE.md` the header should read `## 1.21.1 in progress` rather than a pinned date — the concrete date goes in only once `VERSION` drops the `-rc.X` suffix for the stable release. The last three rc.0s all did this (`## 1.21.0 in progress`, …). `2026-06-01` is also a future date that can't be the final date for an rc that still needs the 3-day prod soak. Using `in progress` also drops the ` / ` separator, which the recent headers no longer use.

File: pkg/cortexpb/timeseriesv2_test.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ func TestTimeseriesV2FromPool(t *testing.T) {
 		ts.LabelsRefs = []uint32{1, 2}
 		ts.Samples = []Sample{{Value: 1, TimestampMs: 2}}
 		ts.Exemplars = []ExemplarV2{{LabelsRefs: []uint32{1, 2}, Value: 1, Timestamp: 2}}
-		ts.Histograms = []Histogram{{}}
+		ts.Histograms = []WrappedHistogram{{}}
```
Comment (sandy2008) [inline pkg/cortexpb/timeseriesv2_test.go:45]: One nit on the conflict resolution: the PR description says the `CreatedTimestamp`/`Metadata` lines were dropped because those fields "don't exist on 1.21", but they do — `TimeSeriesV2.CreatedTimestamp`, `.Metadata`, and `type MetadataV2` are all present on `release-1.21`. The real reason applies only to `CreatedTimestamp`: `ReuseTimeseriesV2` here doesn't clear it, so upstream's `assert.Zero(reused.CreatedTimestamp)` would have failed.

The `Metadata` half, though, `ReuseTimeseriesV2` *does* reset (`Metadata.{Type,UnitRef,HelpRef}`), so dropping it loses real reuse-clearing coverage that 1.21 has. Might be worth re-adding `ts.Metadata = MetadataV2{Type: 1, HelpRef: 2, UnitRef: 3}` before the reuse and asserting it's zeroed afterwards (and tweaking the rationale in the description).

File: pkg/util/http_test.go [hunk 5]
```diff
@@ -220,3 +221,26 @@ func TestIsRequestBodyTooLargeRegression(t *testing.T) {
...
+	// The compressed payload is small enough to pass the compressed-size limit,
+	// but decompresses to far more than maxSize.
+	require.Less(t, compressed.Len(), maxSize)
+
+	var fromWire cortexpb.PreallocWriteRequest
+	err = util.ParseProtoReader(context.Background(), io.NopCloser(&compressed), 0, maxSize, &fromWire, util.Gzip)
+	// The decompressed output should be limited to maxSize+1 bytes, causing a
+	// proto unmarshal error (not an OOM). The key assertion is that we don't
+	// allocate 1 MB of memory.
+	assert.NotNil(t, err)
```
Comment (sandy2008) [inline pkg/util/http_test.go:245]: This test passes with or without the fix, so it doesn't really exercise the new cap. On the unpatched path the gzip branch still reads the whole 1 MB into the buffer and then the pre-existing `len(body) > maxSize` check returns an error → `assert.NotNil(err)` is satisfied either way (and nothing here measures the allocation the comment mentions). The table test above has no gzip case either.

To make it bite, assert the specific size-limit error (`messageSizeLargerErrFmt` / `errors.Is`), or wrap the source in a byte-counting reader and assert at most `maxSize+1` bytes are pulled from the gzip reader. (Carryover from #7515.)

File: pkg/cortex/modules_test.go [hunk 6]
```diff
@@ -248,3 +252,106 @@ func Test_initResourceMonitor_shouldFailOnInvalidResource(t *testing.T) {
...
+			setAllSecrets(v.Elem(), sentinel)
+		}
+	case reflect.Struct:
+		secretType := reflect.TypeFor[flagext.Secret]()
+		for i := 0; i < v.NumField(); i++ {
+			f := v.Field(i)
+			if !f.CanSet() {
+				continue
+			}
+			if f.Type() == secretType {
```
Comment (sandy2008) [inline pkg/cortex/modules_test.go:350]: nit: `setAllSecrets` only stamps the sentinel into fields whose type is `flagext.Secret`, so it skips the one genuinely new secret this release adds — `SignWriteRequestsKeys flagext.SecretStringSliceCSV`. Combined with `sensitivePattern` above not matching `sign_write_requests_keys`, neither guard actually covers the new HMAC keys on `/config`. They *are* masked in practice (`SecretStringSliceCSV.MarshalYAML` → `********`), so no live leak — just a coverage hole. Could teach `setAllSecrets` about `SecretStringSliceCSV` (via its `Set`) so a regression to cleartext would be caught.

File: integration/grpc_compression_test.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+			require.NoError(t, err)
+			require.Equal(t, 200, res.StatusCode)
+
+			res, err = c.Push(series2)
+			require.NoError(t, err)
+			require.Equal(t, 200, res.StatusCode)
+
+			// Wait until the TSDB head is shipped to storage and the store-gateway picks it up.
+			require.NoError(t, ingester.WaitSumMetrics(e2e.Greater(0), "cortex_ingester_shipper_uploads_total"))
+			require.NoError(t, storeGateway.WaitSumMetrics(e2e.Greater(0), "cortex_storegateway_bucket_sync_total"))
```
Comment (sandy2008) [inline integration/grpc_compression_test.go:119]: Question on this subtest: the ingester stays up (`StartAndWaitReady(distributor, ingester, storeGateway)`) and the only gate before the query is `cortex_storegateway_bucket_sync_total > 0`, which just means a sync cycle ran — it doesn't guarantee the block is loaded or that `c.Query("series_1")` is served by the store-gateway rather than the still-live ingester. Does this reliably go through the store-gateway gRPC client (the thing the compression flag actually applies to)? Might be safer to stop the ingester before querying, and/or assert a store-gateway query metric moved. (Carryover from #7459.)

File: pkg/util/flagext/secretstringslicecsv.go (hunk 1 above)
Comment (friedrichg) [reply pkg/util/flagext/secretstringslicecsv.go:23]: https://github.com/cortexproject/cortex/issues/7579

File: integration/ingester_stream_push_test.go (hunk 2 above)
Comment (friedrichg) [reply integration/ingester_stream_push_test.go:207]: https://github.com/cortexproject/cortex/issues/7580

File: integration/grpc_compression_test.go (hunk 7 above)
Comment (friedrichg) [reply integration/grpc_compression_test.go:119]: https://github.com/cortexproject/cortex/issues/7583

File: pkg/cortex/modules_test.go (hunk 6 above)
Comment (friedrichg) [reply pkg/cortex/modules_test.go:350]: https://github.com/cortexproject/cortex/issues/7582

File: pkg/util/http_test.go (hunk 5 above)
Comment (friedrichg) [reply pkg/util/http_test.go:245]: https://github.com/cortexproject/cortex/issues/7581

File: CHANGELOG.md (hunk 3 above)
Comment (friedrichg) [reply CHANGELOG.md]: Done in 6ce3ad75bb — header is now `## 1.21.1 in progress` (no date, no `/` separator), matching the rc.0 convention from the previous releases.

File: pkg/cortexpb/timeseriesv2_test.go (hunk 4 above)
Comment (friedrichg) [reply pkg/cortexpb/timeseriesv2_test.go:45]: Good catch — you're right that `MetadataV2` exists on release-1.21 and is reset by `ReuseTimeseriesV2`. Fixed in 2866b213b5: re-added the `ts.Metadata = MetadataV2{Type: 1, HelpRef: 2, UnitRef: 3}` setup line and the three `assert.Zero` checks for `Metadata.{Type,HelpRef,UnitRef}`. `CreatedTimestamp` stays dropped since `ReuseTimeseriesV2` on 1.21 doesn't clear it (the upstream `assert.Zero` would fail). PR description has been corrected too.

**Review by sandy2008 — APPROVED** [review]
LGTM!

**Review by SungJin1212 — APPROVED** [review]
thanks
